go/vt/vtgate/planbuilder/operators/operator_funcs.go

```diff
@@ -368,3 +299,104 @@ func BreakExpressionInLHSandRHS(
 	ctx.JoinPredicates[expr] = append(ctx.JoinPredicates[expr], rewrittenExpr)
 	return
 }
+
+type joinOperator interface {
+	Operator
+	getLHS() Operator
+	getRHS() Operator
+	setLHS(Operator)
+	setRHS(Operator)
+	makeInner()
+	isInner() bool
+	addJoinPredicate(ctx *plancontext.PlanningContext, expr sqlparser.Expr) error
+}
+
+func addPredicate(join joinOperator, ctx *plancontext.PlanningContext, expr sqlparser.Expr) (Operator, error) {
+	deps := ctx.SemTable.RecursiveDeps(expr)
+	switch {
+	case deps.IsSolvedBy(TableID(join.getLHS())):
+		// predicates can always safely be pushed down to the lhs if that is all they depend on
+		lhs, err := PushPredicate(ctx, expr, join.getLHS())
+		if err != nil {
+			return nil, err
+		}
+		join.setLHS(lhs)
+		return join, err
+	case deps.IsSolvedBy(TableID(join.getRHS())):
+		// if we are dealing with an outer join, always start by checking if this predicate can turn
+		// the join into an inner join
+		if !join.isInner() && canConvertToInner(ctx, expr, TableID(join.getRHS())) {
+			join.makeInner()
+		}
+
+		if !join.isInner() {
+			// if we still are dealing with an outer join
+			// we need to filter after the join has been evaluated
+			return newFilter(join, expr), nil
+		}
+
+		// For inner joins, we can just push the filtering on the RHS
+		rhs, err := PushPredicate(ctx, expr, join.getRHS())
+		if err != nil {
+			return nil, err
+		}
+		join.setRHS(rhs)
+		return join, err
+
+	case deps.IsSolvedBy(TableID(join)):
+		// if we are dealing with an outer join, always start by checking if this predicate can turn
+		// the join into an inner join
+		if !join.isInner() && canConvertToInner(ctx, expr, TableID(join.getRHS())) {
+			join.makeInner()
+		}
+
+		if !join.isInner() {
```

Comment:
it seems that `makeInner` always change the join into an inner join for both `applyJoin` and `join`.